stl/inc/charconv

```diff
@@ -1283,36 +1347,36 @@ _NODISCARD errc _Assemble_floating_point_value(const uint64_t _Initial_mantissa,
     const int32_t _Normal_mantissa_shift  = static_cast<int32_t>(_Traits::_Mantissa_bits - _Initial_mantissa_bits);
     const int32_t _Normal_exponent        = _Initial_exponent - _Normal_mantissa_shift;
 
-    uint64_t _Mantissa = _Initial_mantissa;
-    int32_t _Exponent  = _Normal_exponent;
-
     if (_Normal_exponent > _Traits::_Maximum_binary_exponent) {
         // The exponent is too large to be represented by the floating-point type; report the overflow condition:
         _Assemble_floating_point_infinity(_Is_negative, _Result);
         return errc::result_out_of_range; // Overflow example: "1e+1000"
     }
 
+    uint64_t _Mantissa = _Initial_mantissa;
+    int32_t _Exponent  = _Normal_exponent;
+    errc _Error_code{};
+
     if (_Normal_exponent < _Traits::_Minimum_binary_exponent) {
         // The exponent is too small to be represented by the floating-point type as a normal value, but it may be
-        // representable as a denormal value. Compute the number of bits by which we need to shift the mantissa
-        // in order to form a denormal number. (The subtraction of an extra 1 is to account for the hidden bit of
-        // the mantissa that is not available for use when representing a denormal.)
-        const int32_t _Denormal_mantissa_shift =
-            _Normal_mantissa_shift + _Normal_exponent + _Traits::_Exponent_bias - 1;
+        // representable as a denormal value.
+
+        // The exponent of subnormal values (as defined by the mathematical model of floating point numbers, not the
+        // exponent field in the bit representation) is equal to the minimum exponent of normal values.
+        _Exponent = _Traits::_Minimum_binary_exponent;
 
-        // Denormal values have an exponent of zero, so the debiased exponent is the negation of the exponent bias:
-        _Exponent = -_Traits::_Exponent_bias;
+        // Compute the number of bits by which we need to shift the mantissa in order to form a denormal number.
+        const int32_t _Denormal_mantissa_shift = _Initial_exponent - _Exponent;
 
         if (_Denormal_mantissa_shift < 0) {
-            // Use two steps for right shifts: for a shift of N bits, we first shift by N-1 bits,
-            // then shift the last bit and use its value to round the mantissa.
             _Mantissa =
                 _Right_shift_with_rounding(_Mantissa, static_cast<uint32_t>(-_Denormal_mantissa_shift), _Has_zero_tail);
 
-            // If the mantissa is now zero, we have underflowed:
+            // TRANSITION, existing:
+            // from_chars in MSVC STL and strto[f|d|ld] in UCRT reports underflow only when the result is zero after
+            // rounding to the floating point format. This behavior is different from IEEE 754 underflow exception.

```

Comment:
Is this a correctness issue that we should file in GitHub for the STL? ("TRANSITION" comments are our polite way of saying "BUG", for compiler/STL/etc. bugs.) If it's simply a difference in interpretation of "underflow" between `from_chars/strtod` and IEEE mathematical functions, I think we can note that, but it shouldn't be "TRANSITION". (According to my understanding, which could definitely be wrong, IEEE underflow occurs when a mathematical function needs to return a subnormal value, and it's considered to be a somewhat problematic condition because subnormals effectively lose a lot of precision, and they can cause FPUs to run slower. But for the purposes of `charconv/strtod/etc.`, subnormals are just a slightly different encoding. When charconv talks about underflow, it means that it can't represent the value at all, even as a subnormal. This doesn't seem like a bug, if I'm understanding it correctly.

Reply:
I don't have a strong opinion on whether this should be considered a correctness issue.

Arguments for bug:
- Consistent with IEEE 754 underflow and the notes on `strtod` underflow in the C standard.
- When it is used to parse _a decimal input value_, `from_chars/strtod("1e-323")` returns `9.8813...e-324`, where there _is_ significant loss of precision.

Arguments for non-bug:
- When it is used to parse _the decimal representation of a binary floating-point value_, it round-trips.
- `istream`/`stod`/etc. fails when `strtod` reports underflow, making them not useful for parsing subnormals if `strtod` reports underflow. There're [lots](https://stackoverflow.com/questions/48086830/) [of](https://stackoverflow.com/questions/52410931/) [complaints](https://stackoverflow.com/questions/54652243/) about this.
   This issue is less severe for `from_chars`, which has different return values for format errors and range errors (but there's also LWG-3081).

Reply:
Thanks for the explanation - I like the current form of your comment.